### PR TITLE
Raycast.wurst

### PR DESCRIPTION
**Pull request by GetLocalPlayer**
Contains

For 2D:
- [x] point projection onto ray
- [x] ray X line
- [x] ray X ray
- [x] ray X segment
- [x] ray X circle 

For 3D:
- [x] point projection onto ray
- [x] ray X plane
- [x] ray X disc
- [x] ray X sphere
- [x] ray X triangle

Of course, there are lots of extra things that can be implemented but they can be surplus in Wc3 for stdlib.

### CODE REVIEW

**Comment by Frotty** [comment]
The 3d functions look a bit messy, but other than that good stuff, though on the brim for stdlib.
You should recheck doc for typos "if it behind" and add some package doc where you explain the result tuple and perhaps a small wc3 use case.
Sry for taking some time to check up.

**Comment by GetLocalPlayer** [comment]
> add some package doc where you explain that functions return the result tuple 

Can I add the hotdoc before `package Raycast` at the top of the file?

> though on the brim for stdlib.

There are several functions that I haven't dared to add

For 2D:
- ray X rectangle
- ray X convex polygon

For 3D:
- ray X box
- ray X cylinder (cone)

Do these fit stdlib? Are these overkill?

**Comment by Frotty** [comment]
> Can I add the hotdoc above package Raycast at the top of the file?

I prefer it after the package and import directives. You can use normal /* */

> Do these fit stdlib? Are these overkill?

Probably overkill, also depends on size of course.

**Comment by GetLocalPlayer** [comment]
There's some [map example](https://github.com/GetLocalPlayer/RaycastExample) if it could be useful. Dunno why.

File: wurst/math/Raycast.wurst [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+backface - `true` if the surface at the intersection point is faced away from the ray's origin  */
+public tuple result3d(bool success, vec3 point, real distance, bool backface)
+
+/** Double 3D intersection result.
+'first' is nearest to the ray's origin. */
+public tuple doubleresult3d(result3d first, result3d second)
+
+/* =============== 2D =============== */
+
+function vec2.cross(vec2 v) returns real
```
Comment (Frotty) [inline wurst/math/Raycast.wurst]: why is this here? Belongs into vector

File: wurst/math/Raycast.wurst [hunk 2]
```diff
@@ -0,0 +1,483 @@
+package Raycast
+import NoWurst
+import Vectors
+import Wurstunit
+
+/* The package defines such geometry as 'ray' for 2D and 3D separately.
```
Comment (Frotty) [inline wurst/math/Raycast.wurst]: You should add a small code snippet of one of the raycasts in practical use, showing how to handle the returned tuple correctly.

File: wurst/math/Raycast.wurst (hunk 1 above)
Comment (GetLocalPlayer) [reply wurst/math/Raycast.wurst]: Because there's no cross product of two 2d vectors. This is a simplification for several functions within the package.

**Comment by GetLocalPlayer** [comment]
ain't sure how it can be incorrect. Did you mean that

**Comment by Frotty** [comment]
Yea that's good. Last thing I'm thinking about is whether "success" is the correct wording, since nothing really succeeds here. `if ray.intersects` sounds better than `if ray.success` to me, so that might be better?

**Comment by GetLocalPlayer** [comment]
I think you're right. `success` would be better for functions like `.intesectSomething` but not for raycasting where any result is basically succeed.

**Comment by Frotty** [comment]
I think except for spaces-tabs issue this looks good now 👍

**Comment by GetLocalPlayer** [comment]
Whitspaces replaced with tabs.
